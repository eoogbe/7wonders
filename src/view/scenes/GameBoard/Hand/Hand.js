/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { compose } from 'ramda';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from 'model/actions';
import { getSelectedCardIndex } from 'model/selectors/app';
import {
  getCanBuildDiscarded, getIsCopyingGuild,
} from 'model/selectors/players';
import { getHand } from 'presentation/hand';
import './Hand.css';
import HandCard from './HandCard';

const renderCard = ({ selectedCardIdx, ...cardActions }) =>
  (card, idx) => (
    <HandCard
      key={`${card.id}${idx}`}
      {...cardActions}
      {...card}
      index={idx}
      isSelected={idx === selectedCardIdx}
    />
  );

const Hand = () => {
  const selectedCardIdx = useSelector(getSelectedCardIndex);
  const cards = useSelector(getHand);
  const canBuildDiscarded = useSelector(getCanBuildDiscarded);
  const isCopyingGuild = useSelector(getIsCopyingGuild);

  const dispatch = useDispatch();
  const showHandCallout = compose(dispatch, actions.showHandCallout);
  const buildStructure = compose(dispatch, actions.buildStructure);
  const copyGuild = compose(dispatch, actions.copyGuild);
  let select = compose(dispatch, actions.selectHandCard);

  if (canBuildDiscarded) {
    select = buildStructure;
  }

  if (isCopyingGuild) {
    select = copyGuild;
  }

  if (!cards.length) {
    return <div className="Hand-emptyCardList"></div>
  }

  return (
    <section>
      {canBuildDiscarded &&
        <h2 className="Hand-heading">
          Select a card from the discard pile to build for free:
        </h2>}
      {isCopyingGuild &&
        <h2 className="Hand-heading">
          Select a neighbor's guild card to copy into your tableau:
        </h2>}
      <ul className="Hand-cardList">
        {cards.map(renderCard({
          selectedCardIdx,
          select,
          showHandCallout,
        }))}
      </ul>
    </section>
  );
};

export default Hand;
