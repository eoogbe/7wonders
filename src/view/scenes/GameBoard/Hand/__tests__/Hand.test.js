/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {
  buildStructure, copyGuild, selectHandCard, showHandCallout,
} from 'model/actions';
import Hand from '../Hand';

jest.useFakeTimers();

const mockStore = configureStore();

it('renders without crashing', () => {
  const store = mockStore(createState({ selectedCardIdx: 0 }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <Hand />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders without cards', () => {
  const store = mockStore(createState({
    selectedCardIdx: null,
    hand: [],
  }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <Hand />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders when can build discarded structure', () => {
  const store = mockStore(createState({
    selectedCardIdx: null,
    wonderAbilities: ['BUILD_DISCARDED'],
  }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <Hand />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders when copying a guild card', () => {
  const store = mockStore(createState({
    selectedCardIdx: 0,
    wonderAbilities: ['COPY_GUILD'],
  }));
});

it('selects a card', () => {
  const store = mockStore(createState({ selectedCardIdx: 1 }));
  const tree = mount(
    <Provider store={store}>
      <Hand />
    </Provider>
  );
  tree.find('.HandCard').first().simulate('click');
  expect(store.getActions()).toEqual([selectHandCard('EAST_TRADING_POST', 0)]);
});

it('builds a discarded structure', () => {
  const store = mockStore(createState({
    selectedCardIdx: null,
    wonderAbilities: ['BUILD_DISCARDED'],
  }));
  const tree = mount(
    <Provider store={store}>
      <Hand />
    </Provider>
  );
  tree.find('.HandCard').first().simulate('click');
  expect(store.getActions()).toEqual([buildStructure('LOOM')]);
});

it('copies a guild card', () => {
  const store = mockStore(createState({
    selectedCardIdx: 0,
    wonderAbilities: ['COPY_GUILD'],
  }));
  const tree = mount(
    <Provider store={store}>
      <Hand />
    </Provider>
  );
  tree.find('.HandCard').first().simulate('click');
  expect(store.getActions()).toEqual([copyGuild('CRAFTSMENS_GUILD')]);
});

it('shows the hand callout', () => {
  const store = mockStore(createState({ selectedCardIdx: 1 }));
  const tree = mount(
    <Provider store={store}>
      <Hand />
    </Provider>
  );
  tree.find('.HandCard').first().simulate('mouseEnter');
  expect(store.getActions()).toEqual([showHandCallout('EAST_TRADING_POST')]);
});

const createState = ({ selectedCardIdx, ...player }) => ({
  app: {
    currentPlayer: 'PLAYER1',
    selectedCardIdx,
  },
  ages: [
    { rounds: 6, discardedCards: [] },
    { rounds: 6, discardedCards: [] },
    { rounds: 6, discardedCards: ['LOOM', 'APOTHECARY', 'BATHS'] },
  ],
  players: {
    PLAYER1: {
      id: 'PLAYER1',
      wonder: 'PYRAMIDS_OF_GIZA',
      wonderSide: 'A',
      completedWonderStageCount: 1,
      hand: ['EAST_TRADING_POST', 'GLASSWORKS', 'MINE'],
      tableau: ['STOCKADE', 'TAVERN', 'GLASSWORKS'],
      money: 3,
      wonderAbilities: [],
      ...player,
    },
    PLAYER2: {
      id: 'PLAYER2',
      wonder: 'COLOSSUS_OF_RHODES',
      tableau: ['CLAY_PIT', 'BUILDERS_GUILD'],
    },
    PLAYER3: {
      id: 'PLAYER3',
      wonder: 'TEMPLE_OF_ARTEMIS_IN_EPHESUS',
      tableau: ['CRAFTSMENS_GUILD', 'WORKSHOP'],
    },
  },
})
