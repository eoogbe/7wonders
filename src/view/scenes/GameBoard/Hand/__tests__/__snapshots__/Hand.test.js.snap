// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when can build discarded structure 1`] = `
<section>
  <h2
    className="Hand-heading"
  >
    Select a card from the discard pile to build for free:
  </h2>
  <ul
    className="Hand-cardList"
  >
    <li
      className="HandCard"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(189,189,189)",
        }
      }
    >
      <ul>
        <li>
          Loom
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="HandCard-name"
        >
          Loom
        </div>
      </div>
    </li>
    <li
      className="HandCard"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(76,175,80)",
        }
      }
    >
      <ul>
        <li>
          Compass
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="HandCard-name"
        >
          Apothecary
        </div>
      </div>
    </li>
    <li
      className="HandCard"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(33,150,243)",
        }
      }
    >
      <ul>
        <li>
          3 points
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="HandCard-name"
        >
          Baths
        </div>
      </div>
    </li>
  </ul>
</section>
`;

exports[`renders without cards 1`] = `
<div
  className="Hand-emptyCardList"
/>
`;

exports[`renders without crashing 1`] = `
<section>
  <ul
    className="Hand-cardList"
  >
    <li
      className="HandCard selected"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(255,235,59)",
        }
      }
    >
      <ul>
        <li>
          R trade
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="HandCard-name"
        >
          East trading post
        </div>
      </div>
    </li>
    <li
      className="HandCard"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(189,189,189)",
        }
      }
    >
      <ul>
        <li>
          Glass
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8 0-1.85.63-3.55 1.69-4.9L16.9 18.31C15.55 19.37 13.85 20 12 20zm6.31-3.1L7.1 5.69C8.45 4.63 10.15 4 12 4c4.42 0 8 3.58 8 8 0 1.85-.63 3.55-1.69 4.9z"
          />
        </svg>
        <div
          className="HandCard-name"
        >
          Glassworks
        </div>
      </div>
    </li>
    <li
      className="HandCard"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(141,110,99)",
        }
      }
    >
      <ul
        className="HandCard-cost"
      >
        <li
          className="HandCard-costItem"
        >
          1 coin
        </li>
      </ul>
      <ul>
        <li>
          Ore / Stone
        </li>
      </ul>
      <div
        className="HandCard-footer"
      >
        <div>
          -
          1
        </div>
        <div
          className="HandCard-name"
        >
          Mine
        </div>
      </div>
    </li>
  </ul>
</section>
`;
